Refactor hipBLAS documentation

amd-jnovotny opened the pull request:
resolves #___

Summary of proposed changes:
- Refactor hipBLAS documentation in accordance with style guide and standard formatting
- Copy edits
- Reorganize directory structure to match docs layout (for example, create how-to directory and move a file there)

amd-jnovotny committed:
Initial draft. Requires more work before push

amd-jnovotny committed:
Refactor remaining files and additional edits

daineAMD commented on docs/how-to/contributing-to-hipblas.rst:
```diff
@@ -0,0 +1,92 @@
+*  Do not break existing test cases.
+*  New unit tests should integrate with the existing GoogleTest framework.
+*  Tests must have good code coverage.
+*  Code must also have benchmark tests, and performance must approach
```
Up for debate but not sure if we need this line. hipBLAS doesn't really contribute to the performance at all. That being said, we obviously don't want to add overhead to the hipblas API over the rocBLAS/cuBLAS API, but adding benchmark tests/saying we require perf to approach theoretical limits here seems misguided.

amd-jnovotny replied:
Sure thing. I'll remove this line (it migh have originally been copied from elsewhere).

daineAMD commented on docs/how-to/contributing-to-hipblas.rst:
```diff
@@ -0,0 +1,92 @@
+
+    /opt/rocm/llvm/bin/clang-format -style=file -i <path-to-source-file>
+
+To format all files, run the following script in rocBLAS directory:
```
Should be hipBLAS not rocBLAS

daineAMD commented on docs/install/Linux_Install_Guide.rst:
```diff
@@ -1,189 +1,118 @@
+A few commands in the script require ``sudo`` access, so it might prompt you for a password.
+
+The install script determines the build platform by querying ``hipconfig --platform``. This value can be explicitly set
+using the environment variables ``HIP_PLATFORM=amd`` or ``HIP_PLATFORM=nvidia``.
```
I think this is better as `variable` since it's only one env-var, but up to you.

amd-jnovotny replied:
I'll reword it so it reads more smoothly.

daineAMD commented on docs/install/Linux_Install_Guide.rst:
```diff
@@ -1,189 +1,118 @@
+Installing and building for Linux
+***********************************
 
-Prerequisites
-=============
-
-* If using the rocBLAS backend on an AMD machine:
```
Maybe should mention the prerequisites page here? Maybe not needed with the actual docs layout, not sure.

amd-jnovotny replied:
Good idea. I'll add it to both this guide and the Windows Install guide too. Will make the changes and push the new commit this morning.

daineAMD reviewed (commented):
Sorry for the delay, some initial comments.

amd-jnovotny committed:
Apply suggestions from team review

TorreZuk commented:
Not sure why other CI was running so added another label, seemed to still generate readthedocs in rocblas... we'll see

daineAMD reviewed (approved):
The warnings about hipblasStatus_t, hipblasComputeType_t, and hipblasOperation_t are because they were moved to the [hipblas-common](https://github.com/ROCm/hipBLAS-common/blob/develop/library/include/hipblas-common.h#L56) repo. AFAIK we don't have any docs for hipblas-common yet, but I suppose documentation should be added somewhere to mention those types being available in hipblas-common. Maybe can mention here that those types live in hipblas-common for now. Looks good otherwise.

amd-jnovotny commented:
@daineAMD: Thanks for the info about the warnings. I'll take a look as how to best handle this and will let you know when I have something. I'm busy with some higher-priority work right now, so it might take a few days.

amd-jnovotny committed:
Fix errors in API reference

amd-jnovotny commented:
@daineAMD: I added "For information about the hipblasStatus_t, hipblasComputeType_t, and hipblasOperation_t enumerations, see the hipblas-common GitHub repository." to the top of the API section and also added references under the relevant individual entries. I'm linking to https://github.com/ROCm/hipBLAS-common/blob/develop/library/include/hipblas-common.h. Please let me know if you would like any changes.

daineAMD commented:
Thanks @amd-jnovotny, looks good to me. Feel free to merge if you have permission, otherwise I can merge it if you're happy with it.

amd-jnovotny commented:
@daineAMD Thanks! I'll just have my manager take a quick look, so I'll hold off until that happens. After that, it looks like I have permission to merge. But I'll let you know when I do so.

lpaoletti commented on docs/index.rst:
```diff
@@ -8,33 +8,47 @@
-hipBLAS is a BLAS marshaling library with multiple supported backends. It sits between the application and a 'worker' BLAS library, marshalling inputs into the backend library and marshalling results back to the application.
-hipBLAS exports an interface that does not require the client to change, regardless of the chosen backend. Currently, it supports `rocBLAS <https://github.com/ROCmSoftwarePlatform/rocBLAS>`_ and `cuBLAS <https://developer.nvidia.com/cublas>`_ as backends.
+hipBLAS is a Basic Linear Algebra Subprograms (BLAS) marshaling library with
+support for multiple backends. It lies between the application and a "worker" BLAS library,
```
Perhaps it should be:
marshaling library that supports multiple backends

or:
marshaling library supporting multiple backends

lpaoletti commented on docs/how-to/contributing-to-hipblas.rst:
```diff
@@ -0,0 +1,90 @@
+.. meta::
+  :description: How to contribute to hipBLAS
+  :keywords: hipBLAS, rocBLAS, BLAS, ROCm, API, Linear Algebra, documentation, contribution
```
```suggestion
  :keywords: hipBLAS, rocBLAS, BLAS, ROCm, API, contribution
```

lpaoletti commented on docs/how-to/contributing-to-hipblas.rst:
```diff
@@ -0,0 +1,90 @@
+Contributing to hipBLAS
+********************************************************************
+
+This topic explains how to contribute to the hipBLAS code base, including coding
+and pull request (PR) guidelines, along with information on static code analysis.
```
```suggestion
This topic explains how to contribute to the hipBLAS code base, including coding
and pull request (PR) guidelines and information about static code analysis.
```

lpaoletti commented on docs/how-to/contributing-to-hipblas.rst:
```diff
@@ -0,0 +1,90 @@
+
+The hipBLAS code contribution guidelines closely follow the `GitHub
+pull requests <https://help.github.com/articles/using-pull-requests/>`__ model.
+The hipBLAS repository follows a workflow which dictates a ``release`` branch, from which releases are cut, and a
```
should the first "which" be a "that"

lpaoletti commented on docs/how-to/contributing-to-hipblas.rst:
```diff
@@ -0,0 +1,90 @@
+The hipBLAS code contribution guidelines closely follow the `GitHub
+pull requests <https://help.github.com/articles/using-pull-requests/>`__ model.
+The hipBLAS repository follows a workflow which dictates a ``release`` branch, from which releases are cut, and a
+``develop`` branch which serves as an integration branch for new code. Pull requests should
```
```suggestion
``develop`` branch, which serves as an integration branch for new code. Pull requests should
```

lpaoletti commented on docs/how-to/contributing-to-hipblas.rst:
```diff
@@ -0,0 +1,90 @@
+Coding guidelines:
+==================
+
+*  Do not use unnamed namespaces inside of header files.
```
```suggestion
*  Don't use unnamed namespaces inside of header files.
```

lpaoletti commented on docs/how-to/contributing-to-hipblas.rst:
```diff
@@ -0,0 +1,90 @@
+      ``static`` ``class`` member functions defined in header files are okay.
+
+*  Use ``static`` for ``constexpr`` ``template`` variables until C++17. After C++17,
+   ``constexpr`` variables become ``inline`` variables, and can therefore be defined in multiple compilation units.
```
```suggestion
   ``constexpr`` variables become ``inline`` variables, and can be defined in multiple compilation units.
```

lpaoletti commented on docs/how-to/contributing-to-hipblas.rst:
```diff
@@ -0,0 +1,90 @@
+-----------
+
+C and C++ code is formatted using ``clang-format``. To run ``clang-format``,
+use the version in the ``/opt/rocm/llvm/bin`` directory. Do not use your
```
```suggestion
use the version in the ``/opt/rocm/llvm/bin`` directory. Don't use your
```

lpaoletti commented on docs/how-to/contributing-to-hipblas.rst:
```diff
@@ -0,0 +1,90 @@
+*  Target the **develop** branch for integration.
+*  Ensure code builds successfully.
+*  Do not break existing test cases.
+*  New unit tests should integrate with the existing GoogleTest framework.
```
Do you think GoogleTest should be hyperlinked?

amd-jnovotny replied:
I don't think people need to know much about GoogleTest to use it, so I'll leave it out.

lpaoletti commented on docs/conceptual/library-source-code-organization.rst:
```diff
@@ -0,0 +1,88 @@
+Library source code organization
+**********************************
+
+The hipBLAS code is split into several parts:
```
```suggestion
The hipBLAS code is split into the following parts:
```

lpaoletti replied:
very subjective comment  - your call

lpaoletti commented on docs/conceptual/library-source-code-organization.rst:
```diff
@@ -0,0 +1,88 @@
+The following utilities support the hipBLAS infrastructure:
+
+*  CMake is used to build and package hipBLAS. There are ``CMakeLists.txt`` files throughout the code.
+*  Doxygen, Breathe, Sphinx, and ReadTheDocs are used to generate the documentation. The following sources supply
```
```suggestion
*  Doxygen, Breathe, Sphinx, and ReadTheDocs generate the documentation. The following sources supply
```

lpaoletti commented on docs/install/prerequisites.rst:
```diff
@@ -0,0 +1,44 @@
+.. meta::
+  :description: hipBLAS installation prerequisites for Linux and Windows
+  :keywords: hipBLAS, rocBLAS, BLAS, ROCm, API, Linear Algebra, documentation, install, prerequisites
```
```suggestion
  :keywords: hipBLAS, rocBLAS, BLAS, ROCm, API, Linear Algebra, install, prerequisites
```

lpaoletti commented on docs/install/prerequisites.rst:
```diff
@@ -0,0 +1,44 @@
+Prerequisites for hipBLAS installation
+***********************************************************
+
+The following prerequisites are required to install hipBLAS, whether you are using a package manager
```
```suggestion
The following prerequisites are required to install hipBLAS, whether using a package manager
```

lpaoletti commented on docs/install/prerequisites.rst:
```diff
@@ -0,0 +1,44 @@
+Prerequisites for Linux
+=========================
+
+The hipBLAS prerequisites are somewhat different for the :doc:`rocBLAS <rocblas:index>` and NVIDIA CUDA `cuBLAS <https://developer.nvidia.com/cublas>`_ backends.
```
```suggestion
The hipBLAS prerequisites are different for the :doc:`rocBLAS <rocblas:index>` and NVIDIA CUDA `cuBLAS <https://developer.nvidia.com/cublas>`_ backends.
```

lpaoletti replied:
Find this sentence awkward. 
Maybe this:
The hipBLAS prerequisites are different than the [rocBLAS](https://rocm.docs.amd.com/projects/rocBLAS/en/latest/index.html) and NVIDIA CUDA [cuBLAS](https://developer.nvidia.com/cublas) backend prerequisites.

lpaoletti commented on docs/install/prerequisites.rst:
```diff
@@ -0,0 +1,44 @@
+*  Here are the prerequisites required to use the rocBLAS backend with AMD components:
+
+   * A ROCm-enabled platform. For more information, see the :doc:`Linux system requirements <rocm-install-on-linux:reference/system-requirements>`.
+   * A compatible version of hipblas-common.
```
hyperlink here?

lpaoletti commented on docs/install/prerequisites.rst:
```diff
@@ -0,0 +1,44 @@
+   * A ROCm-enabled platform. For more information, see the :doc:`Linux system requirements <rocm-install-on-linux:reference/system-requirements>`.
+   * A compatible version of hipblas-common.
+   * A compatible version of rocBLAS.
+   * For full functionality, a compatible version of :doc:`rocSOLVER <rocsolver:index>` and its :doc:`rocSPARSE <rocsparse:index>`
```
the term "full functionality" is interesting.
What happens if you don't have this. Sounds optional.

lpaoletti commented on docs/install/prerequisites.rst:
```diff
@@ -0,0 +1,44 @@
+   * For full functionality, a compatible version of :doc:`rocSOLVER <rocsolver:index>` and its :doc:`rocSPARSE <rocsparse:index>`
+     and :doc:`rocPRIM <rocprim:index>` dependencies.
+
+*  Here are the prerequisites required to use the cuBLAS backend with NVIDIA components:
```
```suggestion
*  The prerequisites required to use the cuBLAS backend with NVIDIA components:
```

lpaoletti replied:
not sure. 
if you change this, change the above as well.

lpaoletti commented on docs/install/prerequisites.rst:
```diff
@@ -0,0 +1,44 @@
+Prerequisites for Windows
+=========================
+
+*  Here are the prerequisites required to use the rocBLAS backend with AMD components:
```
Similar comments that were made above in Linux might apply to Windows.

lpaoletti commented on docs/install/prerequisites.rst:
```diff
@@ -0,0 +1,44 @@
+   * A compatible version of rocBLAS.
+   * For full functionality, a compatible version of :doc:`rocSOLVER <rocsolver:index>` and its :doc:`rocSPARSE <rocsparse:index>`
+     and :doc:`rocPRIM <rocprim:index>` dependencies.
+   * hipBLAS is supported on the same Windows versions and toolchains that the HIP SDK supports.
```
```suggestion
   * hipBLAS is supported on the same Windows versions and toolchains that HIP SDK supports.
```

lpaoletti commented on docs/install/prerequisites.rst:
```diff
@@ -0,0 +1,44 @@
+     and :doc:`rocPRIM <rocprim:index>` dependencies.
+   * hipBLAS is supported on the same Windows versions and toolchains that the HIP SDK supports.
+
+* hipBLAS does not currently support the cuBLAS backend on Windows.
```
```suggestion
* hipBLAS does not support the cuBLAS backend on Windows.
```

lpaoletti commented on docs/install/prerequisites.rst:
```diff
@@ -0,0 +1,44 @@
+   * A HIP-enabled platform. For more information, see :doc:`HIP installation <hip:install/install>`.
+   * A working CUDA toolkit, including cuBLAS. For more information, see `CUDA toolkit <https://developer.nvidia.com/accelerated-computing-toolkit/>`_.
+
+Prerequisites for Windows
```
```suggestion
Prerequisites for Microsoft Windows
```

lpaoletti replied:
first occurence.

TorreZuk replied:
@lpaoletti can you please use the review feature rather than multiple comments?

lpaoletti commented on docs/install/Linux_Install_Guide.rst:
```diff
@@ -1,189 +1,119 @@
-
-  * A HIP enabled platform, more information :doc:`HIP installation <hip:how_to_guides/install>` page.
-  * A working CUDA toolkit, including cuBLAS, see `CUDA toolkit <https://developer.nvidia.com/accelerated-computing-toolkit/>`_.
+This topic discusses how to install hipBLAS from a package and how to build and install it from the source code.
```
Perhaps intro paragraph should have "Linux" somewhere in there.

lpaoletti commented on docs/install/Linux_Install_Guide.rst:
```diff
@@ -1,189 +1,119 @@
+This topic discusses how to install hipBLAS from a package and how to build and install it from the source code.
+For a list of installation prerequisites, see :doc:`hipBLAS prerequisites <prerequisites>`.
 
 Installing pre-built packages
```
```suggestion
Installing prebuilt packages
```

lpaoletti commented on docs/install/Linux_Install_Guide.rst:
```diff
@@ -1,189 +1,119 @@
 =============================
 
-Download pre-built packages either from `ROCm's native package manager <https://rocm.docs.amd.com/projects/install-on-linux/en/latest/tutorial/quick-start.html#native-package-manager>`_ or by clicking the `GitHub releases tab <https://github.com/ROCm/hipBLAS/releases>`_ and manually downloading, which could be newer.  Release notes are available for each release on the releases tab.
+You can manually download the pre-built hipBLAS packages from :doc:`ROCm's native package manager <rocm-install-on-linux:install/quick-start>`
```
```suggestion
You can manually download the pre-built hipBLAS packages from the :doc:`ROCm native package manager <rocm-install-on-linux:install/quick-start>`
```

lpaoletti replied:
Avoid making our trademarks possessive.

lpaoletti commented on docs/install/Linux_Install_Guide.rst:
```diff
@@ -1,189 +1,119 @@
-
-Build library dependencies + client dependencies + library + client
+When building hipBLAS from source, you can choose to build only the library and its dependencies or include the client and its
+dependencies as well.
```
```suggestion
When building hipBLAS from source, you can choose to build only the library and its dependencies or include the client and its
dependencies.
```

lpaoletti commented on docs/install/Linux_Install_Guide.rst:
```diff
@@ -1,189 +1,119 @@
+Building the library and library dependencies
+---------------------------------------------
+
+The root directory of this repository contains the helper Python script ``rmake.py`` that lets you build and install
+hipBLAS with a single command. It accepts several options but has a hard-coded configuration
+that you can override by invoking ``cmake`` directly. However, it's a great way to get started quickly and
+serves as an example of how to build and install hipBLAS.
+A few commands in the script require ``sudo`` access, so it might prompt you for a password.
```
```suggestion
The root directory of this repository contains the helper Python script ``rmake.py`` which lets you build and install
hipBLAS with a single command. It accepts several options but has a hard-coded configuration
that you can override by invoking ``cmake`` directly. However, it's a great way to get started quickly and
serves as an example of how to build and install hipBLAS.
A few commands in the script require ``sudo`` access, which might prompt you for a password.
```

lpaoletti commented on docs/install/Linux_Install_Guide.rst:
```diff
@@ -1,189 +1,119 @@
+   "``python3 rmake.py -d``", "Build the library dependencies and library in your local directory. The ``-d`` flag only has to be used once. For subsequent invocations of ``rmake.py``, it is not necessary to rebuild the dependencies."
+   "``python3 rmake.py``", "Build the library in your local directory. It is assumed the dependencies have been built."
+   "``python3 rmake.py -i``", "Build the library, then build and  install the hipBLAS package in  ``/opt/rocm/hipblas``. You will be prompted for sudo access. This installs it for all users. To restrict hipBLAS to your local directory, do not use the  ``-i`` flag. "
+   "``python3 rmake.py -n``", "Build the library without the functionality provided by rocSOLVER. The rocSOLVER, rocSPARSE, and rocPRIM dependencies will not be required. This flag has no effect when building with a CUDA backend."
```
```suggestion
   "``python3 rmake.py -n``", "Build the library without the functionality provided by rocSOLVER. The rocSOLVER, rocSPARSE, and rocPRIM dependencies will not be required. This flag has no effect when building with a NVIDIA CUDA backend."
```

lpaoletti replied:
First occurence?

lpaoletti commented on docs/install/Linux_Install_Guide.rst:
```diff
@@ -1,189 +1,119 @@
-
-Manual build (all supported platforms)
+================= =======================================================
+hipblas-test      Runs GoogleTest tests to validate the library
```
link to GoogleTest?

lpaoletti commented on docs/install/Linux_Install_Guide.rst:
```diff
@@ -195,24 +125,32 @@ Build Library Using Individual Commands
 
-- `lapack <https://github.com/Reference-LAPACK/lapack-release>`_,  lapack itself brings a dependency on a fortran compiler
-- `googletest <https://github.com/google/googletest>`_
+Unfortunately, GoogleTest and LAPACK are more difficult to install. Many distributions
```
```suggestion
GoogleTest and LAPACK are more difficult to install. Many distributions
```

lpaoletti commented on docs/install/Windows_Install_Guide.rst:
```diff
@@ -1,211 +1,143 @@
-  * hipBLAS is supported on the same Windows versions and toolchains that are supported by the HIP SDK.
-
-* hipBLAS does not currently support using the cuBLAS backend on Windows.
+This topic covers how to install hipBLAS on Windows from a package and how to build and install it from the source code.
```
```suggestion
This topic covers how to install hipBLAS on Microsoft Windows from a package and how to build and install it from the source code.
```

lpaoletti commented on docs/install/Windows_Install_Guide.rst:
```diff
@@ -1,211 +1,143 @@
+This topic covers how to install hipBLAS on Windows from a package and how to build and install it from the source code.
+For a list of installation prerequisites, see :doc:`hipBLAS prerequisites <prerequisites>`.
 
 Installing pre-built packages
```
```suggestion
Installing prebuilt packages
```

lpaoletti replied:
search page for this.

lpaoletti commented on docs/install/Windows_Install_Guide.rst:
```diff
@@ -1,211 +1,143 @@
 =============================
 
-hipBLAS can be installed on Windows 11 or Windows 10 using the AMD HIP SDK installer.
+hipBLAS can be installed on Windows 10 or 11 using the AMD HIP SDK installer.
```
link to HIP SDK?

lpaoletti commented on docs/install/Windows_Install_Guide.rst:
```diff
@@ -1,211 +1,143 @@
 
-The hipBLAS source code is available at the `hipBLAS github page <https://github.com/ROCm/hipBLAS>`_. The version of the ROCm HIP SDK may be shown in the path of default installation, but you can run the HIP SDK compiler to report the versino from the bin/ folder with:
+The hipBLAS source code is available on the `hipBLAS GitHub <https://github.com/ROCm/hipBLAS>`_.
+The version of the ROCm HIP SDK might be shown in the default installation path,
+but you can run the HIP SDK compiler to display the version from the ``bin/`` folder:
```
```suggestion
The ROCm HIP SDK version might be shown in the default installation path,
but you can run the HIP SDK compiler to display the version from the ``bin/`` folder:
```

lpaoletti commented on docs/install/Windows_Install_Guide.rst:
```diff
@@ -1,211 +1,143 @@
+with a single command. It accepts several options but has a hard-coded configuration
+that you can override by invoking ``cmake`` directly. However, it's a great way to get started quickly and serves
+as an example of how to build and install hipBLAS.
+A few commands in the script require administrator access, so it might prompt you for a password.
```
```suggestion
A few commands in the script require administrator access, so you might be prompted for a password.
```

lpaoletti replied:
not sure if any better

lpaoletti commented on docs/how-to/using-hipblas-clients.rst:
```diff
@@ -0,0 +1,121 @@
+   transA,transB,M,N,K,alpha,lda,ldb,beta,ldc,hipblas-Gflops,us
+   N,N,4096,4096,4096,1,4096,4096,0,4096,11941.5,11509.4
+
+A useful way of finding the parameters that can be used with ``./hipblas-bench -f gemm`` is to turn on logging
```
```suggestion
A helpful way of finding the parameters that can be used with ``./hipblas-bench -f gemm`` is to turn on logging
```

lpaoletti commented on docs/how-to/using-hipblas-clients.rst:
```diff
@@ -0,0 +1,121 @@
+
+   hipblas-bench also provides the flag ``-v 1`` for correctness checks.
+
+If multiple arguments or functions need to be benchmarked,
+hipblas-bench provides support for data-driven benchmarks using a YAML-format specification file.
```
```suggestion
If multiple arguments or functions need to be benchmarked,
hipblas-bench supports data-driven benchmarks using a YAML-format specification file.
```

lpaoletti commented on docs/how-to/using-hipblas-clients.rst:
```diff
@@ -0,0 +1,121 @@
+
+   ./hipblas-bench --yaml <file>.yaml
+
+As an example, ``hipblas_smoke.yaml`` is a YAML file that is used to run a smoke test.
+However, other examples can be found in the `rocBLAS <https://github.com/ROCm/rocBLAS>`_ GitHub repository.
```
```suggestion
For example, ``hipblas_smoke.yaml`` is a YAML file used to run a smoke test.
However, other examples can be found in the `rocBLAS <https://github.com/ROCm/rocBLAS>`_ GitHub repository.
```

lpaoletti commented on docs/reference/deprecation.rst:
```diff
@@ -82,58 +85,60 @@ The above is replaced by an in-place and out-of-place trmm that calculates C <-
+* function ``hipblasSetInt8Datatype`` is removed
+* function ``hipblasGetInt8Datatype`` is removed
 
 Announced in hipBLAS 2.0
```
To confirm, these were announced in 2.0 but none have been actioned yet?

amd-jnovotny committed:
Incorporate comments from review

amd-jnovotny committed:
A few additional edits
